### DISCUSSION TITLE
ESP32-H2 compatibility

### DISCUSSION THREAD

**Discussion by rroblik** [fredlcore/BSB-LAN#691] [General]
Hi, 

I has a lot of freeze issue with my arduino board. I'm looking to upgrade it to a more home automation "compliant" protocol from WiFi to zigbee. I found [esp32-h2](https://www.espressif.com/en/products/socs/esp32-h2) board, so I ask to myself if this can work out of the box with BSB-LAN ? Or what changes will that require? Found nothing about it...

Thanks !

**Comment by fredlcore** [comment]
I don't know - could work, but no guarantee, and no support from my side. A normal ESP32 is cheap, and the boards for right in, so I don't see the need for other models.

**Comment by rroblik** [comment]
Because heating system are commonly far away from the living where the main of the home automation live. Also because I want to stop using WiFi for automation and because I have a lot of zigbee routers around my heating system. What is the main thing to check for compatibility ?

thanks

**Comment by janstuyck** [comment]
Hi rroblik,

According to my experience there is no problem compiling (I found it very easy and little work) for other ESP32 board as long as they have >=4MB flash. At this time I am using a XIAO ESP32-C6 with 4MB flash (I would have preferred 8MB regarding the partition scheme).

The only things I had to do was:

- Change the TX/RX pins in de code to the ones I chose to use on my C6 board (Next to GND and 3.3V these are the four connections that need to be wired to the BSB-LAN adapter you have. My adapter is currently still just connected by four wires...

- I also had to make a custom partitions.csv because for the XIAO_ESP32C6 board theer is no option for a "Minimal SPIFFS" partition scheme as instructed in step 9 of the Quickstart instruction in the manual (https://docs.bsb-lan.de/quickstart.html).

Is it correct that the H2 doesn't come with WiFi? I would suggest you to go for a C6 version because then you can easily start without having to get all the changes for merging from WiFi to ZigBee right. You could leave all the WiFi related code where it is and use it as a reference for your new ZigBee code?

I'm sharing a link to my XIAO ESP32 C6 project. I didn't make myself familiar with working in a Github fork so it's in another cloud:
https://e.pcloud.link/publink/show?code=kZUMQwZXesjqnOqUNJlfRoQmH6rIBlFSxpy

I am also thinking about using its ZigBee / Thread capabilities in some future. Since this not incorporated (yet, will it ever be?) in ESPHome this could be a nice project to familiarize myself with it in the Arduino IDE. So I'm very interested in your work concerning implementation of ZigBee!!

Have fun!

Kind regards,

Jan

**Reply by janstuyck** [reply]
Hi @fredlcore,
Thanks for moving this to a discussion. You're right. And I was not aware it was not a discussion already.. :-)

@rrobllic,
I just thought about the fact that I didn't add comments to everything I changed from the beginning. Most of what I did has somewhere a comment starting with //JS// or so (JS being my initials).
I am reviewing and completing this right now and will let you know when I finished because that will make it easier for you and myself in the future.
What I do remember that is also important to mention: My version compiles thinking an ESP32_NODEMCU board is used. I didn't do the work of making definitions for my XIAO board, but changed the settings for the ESP32_NODEMCU instead, which is of course not exactly nice nor professional of me.. Sorry for that.

**Reply by janstuyck** [reply]
Hi @rroblik,

I finished commenting everything I changed in all of the project in the BSB-LAN folder that automatically open in the Arduino IDE if you open the BSB_LAN.ino project. They all have `//*JS*//` in the comments.

I installed the boards into Ardiono IDE that contains my Seeed Xiao C6 board and select that bord as instructed by Seeed:
https://wiki.seeedstudio.com/xiao_esp32c6_getting_started/
The code detects a ESP32_NODEMCU and compiles with those definitions. I changed the TX2 and RX2 pins to those exposed on my C6 board where they were configured for the ESP32_NODEMCU as you will see.
And as I mentioned I made a custom partitions.csv.

The "Device-specific parameter list" for my ELCO Thision S9 is in the project already.

Greetings,

Jan

**Comment by fredlcore** [comment]
@rroblik: So you mean you want to replace WiFi/Ethernet with ZigBee? That would probably include major changes in the way BSB-LAN works, including the web interface, MQTT and telnet access. I haven't really worked with ZigBee yet, and if the added footprint is tolerable, I would consider adding it sometime in the future, but not as a replacement for networking functionality. 

I'm converting this issue into a discussion, so you can continue discussing the matter without it being on my "to do list".

**Comment by rroblik** [comment]
To be honest I've totally forgotten about the web interface 🙄. Once configured we only look the mqtt playload but's it's true, the web interface exists ! 
@janstuyck i will read all that, seems interesting !

**Reply by fredlcore** [reply]
But how would you use MQTT over ZigBee? Or would you replace MQTT with ZigBee in terms of communication? This sounds convenient once everything is up and running, but in terms of troubleshooting it seems not like a good idea to replace WiFi/LAN completely...

**Reply by janstuyck** [reply]
There exist things like https://www.zigbee2mqtt.io/ so I would'be surprised that both something like MQTT over Zigbee can be implemented next to the existing MQTT over Wifi. Or am I talking nonsense here? :-)

**Reply by fredlcore** [reply]
As far as I understand zigbee2mqtt, this does not run on the individual "client" (such as the ESP32 that runs BSB-LAN), but on some kind of "relay". So BSB-LAN would still just implement ZigBee for exchanging parameter information etc. However, I don't see the benefit if you need to have some kind of network connection anyways (for OTA updates, changing settings, accessing debug messages etc.), why then just send the parameter data over a different kind of network? I see a reason for ZigBee for simple sensors or switches, but rather not for something more complex like BSB-LAN.

**Comment by rroblik** [comment]
Yes you're totally right @fredlcore. I realize what I imagine is not realist !

thanks I'm closing this discussion right now!

**Reply by fredlcore** [reply]
Well, maybe @janstuyck has some reasons for doing the ZigBee implementation that I have overlooked, but unless you completely replace the network stack with ZigBee and fall back on configuring via BSB_LAN_config.h and can live without debug output as well as OTA, it could be possible. But I don't really see a benefit that would outweigh the losses that come with it.

**Reply by janstuyck** [reply]
As @fredlcore says, some essential and a lot of the beautiful extra features that the project offers are out of the scope of ZigBee. Matter/Thread might? I hardly know anything about that and it's probably too soon to find enough people who want to work on that.

Indeed there is no need to discuss about replacing anything network related. It's too beautiful in the first place. Many other reasons. For instance I learned from @fredlcore hat many users and co-developers are not using it to integrate into their Home Assistant or alike domotics built on a server.

If there would be enough manpower to work on something like this it has to be optional to use it and work next to what is. The plus for my particular case is that in my Home Assistant playground I am going for ZigBee devices where possible because more reliable: The coordinator is plugged into the server and ZigBee devices stay connected when WiFi is down. A spare coordinator costs like €20 or so. Also keeping a spare redundant-ich HA server can be cheap. One of the (was it three?) main rules for a smart home to be smart is that most of it still works when a server or WiFi stops working. So for me it would be a nice option. Far from essential. It would at least make myself a lot smarter if alone I would dig into what is already developed and understand half of it :-). But I'm afraid I don't have time to do this by myself or engage myself to help coordinate such a side project that could be added to the big thing without affecting anyone who just wants working updates on what they have and that just works or anyone working on the project the way they do now.

Sorry for the big amount of text. I hope that maybe my previous posts helped @rroblik a little bit. Probably most people who want to use a different microcontroller dev board don't need anything from what I was telling in my previous posts.. :)

**Reply by fredlcore** [reply]
No, thank you for that, I think it's good to understand the pros and cons fully before moving on, and some of the things are certainly important/useful. For example, I would have never guessed how much people would appreciate the MQTT-HA-Autodiscovery feature, simply because I don't use HA. Maybe it's the same with ZigBee, I don't know. MQTT was relatively easy because, initially, we just sent the same strings to MQTT that we would have sent via HTTP. Once such a rather easy add-on was possible, we could continue working from there. I have no idea how complex adding ZigBee is in the first place, but I'll keep it in the back of my head...
